content/posts/wasm-as-a-platform-for-abstraction.md

```diff
@@ -0,0 +1,1390 @@
+
+This can be tricky because we're interacting with a lot of `unsafe`
+code, in particular:
+
+- We can't make the `poll()` function generic over any type `E: Environment`
+  because then when `Ctx::data` is read by our functions, they won't know
+  which type of `*mut E` to cast it to. The easiest way around this is to use
+  dynamic dispatch (i.e. `&mut dyn Environment`)
+- You can't cast a fat pointer (`&mut dyn Environment`) to a thin pointer
+  (`*mut c_void`) so we need a second level of indirection
+- The item pointed to by `Ctx::data` (our `Environment` object) is only
+  guaranteed to stay valid for the duration of `poll()` so we need to make sure
+  it gets cleared before returning
+- Code that we call may `panic!()` and we need to make sure `Ctx::data` is
+  cleared *no matter what*, otherwise we'll be leaving a dangling pointer behind
+  and future calls may try to use it
+
+To solve the first two problems we'll introduce an intermediate `State` object
+which can be placed on the stack.
+
+```rust
+// src/lib.rs
+
+/// Temporary state passed to each host function via [`Ctx::data`].
+struct State<'a> {
+    env: &'a mut dyn Environment,
+}
+```
+
+From here, the naive implementation for `poll()` would look something like this:
+
+```rust
+// src/lib.rs
+
+impl Program {
+    ...
+
+    pub fn poll(&mut self, env: &mut dyn Environment) -> Result<(), Error> {
+        let mut state = State { env };
+        self.instance.context_mut().data = &mut state as *mut State as *mut _;
+
+        self.instance.call("poll", &[])?;
+
+        Ok(())
+    }
+}
+```
+
+And yes, while it *does* correctly thread our `&mut dyn Environment` through to
+the instance-global context data, we've completely ignored the last two points;
+preventing our temporary `state` pointer from dangling, even if `wasmer`
+panics.
+
+The normal way to implement this is by putting `self.instance.call("poll", &[])`
+inside a closure, then using a helper function to
+
+1. Do some setup
+2. Call the closure from `std::panic::catch_unwind()`
+3. Do safety-critical cleanup, then
+4. Resume panicking
+
+```rust
+// src/lib.rs
+
+impl Program {
+    ...
+
+    pub fn poll(&mut self, env: &mut dyn Environment) -> Result<(), Error> {
+        let mut state = State { env };
+        self.instance.context_mut().data = &mut state as *mut State as *mut _;
+
+        self.with_environment_context(env, |instance| {
+            instance.call("poll", &[])?;
+            Ok(())
+        })
+    }
+
+    fn with_environment_context<F, T>(
+        &mut self,
+        env: &mut dyn Environment,
+        func: F,
+    ) -> Result<T, Error>
+    where
+        F: FnOnce(&Instance) -> Result<T, Error>,
+    {
+        let mut state = State { env };
+        let instance = &mut self.instance;
+
+        // point the data pointer at our temporary state.
+        instance.context_mut().data = &mut state as *mut State<'_> as *mut _;
```

Comment:
Is this equivalent of casting to void* in Rust?

Reply:
Yep. The compiler will resolve the `_` in `as *mut _` to `c_void`.